packages/database/src/migrations/1745987213057-OptimiseFhirJobGrabFunction.ts

```diff
@@ -0,0 +1,140 @@
+import { QueryInterface } from 'sequelize';
+
+export async function up(query: QueryInterface): Promise<void> {
+  await query.sequelize.query(`
+    DROP INDEX IF EXISTS fhir."job_grab_idx";
+
+    -- Add index on fhir.jobs table which is optimised for the fhir.job_grab function
+    CREATE INDEX job_grab_idx ON fhir.jobs USING BTREE (topic, status, priority DESC, created_at);
+  `);
+
+  await query.sequelize.query(`
+    -- Rework the fhir.job_grab function to perform 3 separate selects which allows for better use of the new index
+    CREATE OR REPLACE FUNCTION fhir.job_grab(with_worker uuid, from_topic text, OUT job_id uuid, OUT job_payload jsonb)
+      RETURNS record
+      LANGUAGE plpgsql
+      STRICT
+    AS $function$
+    BEGIN
+      IF NOT fhir.job_worker_is_alive(with_worker) THEN
+        RAISE EXCEPTION 'worker % is not alive', with_worker;
+      END IF;
+    
+      -- The next queued job
+      WITH queued_job AS (
+        SELECT id, payload, priority, created_at FROM fhir.jobs
+        WHERE (
+          topic = from_topic
+          AND status = 'Queued'
+        )
+        ORDER BY priority DESC, created_at ASC
+        LIMIT 1
+        FOR UPDATE
+        SKIP LOCKED
+      ),
+      
+      -- A currently grabbed job that hasn't been updated in 10 seconds (it may have gone stale)
+      grabbed_job AS (
+        SELECT id, payload, priority, created_at FROM fhir.jobs
+        WHERE (
+          topic = from_topic
+          AND status = 'Grabbed'
+          AND updated_at < CURRENT_TIMESTAMP - INTERVAL '10 seconds'
+        )
+        ORDER BY priority DESC, created_at ASC
+        LIMIT 1
+        FOR UPDATE
+        SKIP LOCKED
+      ),
+      
+      -- A started job whose worker has died
+      started_job AS (
+        SELECT id, payload, priority, created_at FROM fhir.jobs
+        WHERE (
+          topic = from_topic
+          AND status = 'Started'
+          AND NOT fhir.job_worker_is_alive(worker_id)
+        )
+        ORDER BY priority DESC, created_at ASC
+        LIMIT 1
+        FOR UPDATE
+        SKIP LOCKED
+      )
```

Comment:
Separately selecting the different job candidates like this allows us to use the new index on the `queued_job` which is by far the most common case and greatly improves job grab time when there are a lot of jobs in the queue.

However, there is a slight downside which is that potentially we lock up to 3 jobs while we do this, so other workers cannot grab those as we consider which job to grab. I doubt that's a real issue as those jobs become unlocked as soon as we choose which one to grab and mark it as such.